[java] Merge capabilities of slot with the new session request capabilities

pujagani opened the pull request:
**Thanks for contributing to Selenium!**
**A PR well described will help maintainers to quickly review and merge it**

Before submitting your PR, please check our [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) guidelines.
Avoid large PRs, help reviewers by making them as simple and short as possible.


<!--- Provide a general summary of your changes in the Title above -->

### Description
<!--- Describe your changes in detail -->
Merge capabilities of slot with the new session request capabilities

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Node config allows the Grid user to set driver configuration. 
Example: `java -jar selenium-server-4.6.0.jar standalone --port 50859 --detect-drivers false --driver-configuration display-name="Chrome Custom" stereotype='{"browserName":"chrome","goog:chromeOptions":{"args":[\"incognito\",\"window-size=500,500\"]},"pageLoadStrategy":"normal"}'`

The configuration is applied to the session slots.
This intends to provide the configuration that all matching incoming session requests should ideally consider and pass it along to the respective web driver.
Currently, the new session request capabilities are passed as it is without considering if the session slot has additional capabilities that need to be considered.

The changes help fix this by giving merging the capabilities of the slot and the incoming session request. The capabilities of the slot override the capabilities of the incoming session request in case of custom driver-configuration.


### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.
<!--- Provide a general summary of your changes in the Title above -->

diemol commented:
We do something similar on the [`LocalNode`](https://github.com/SeleniumHQ/selenium/blob/6a8bf1ca7ed2c22eb736abb2ebb5cb58bc5d87bd/java/src/org/openqa/selenium/grid/node/local/LocalNode.java#L538-L539). Does this PR improve that code? Should we centralize that logic and do it in a single place?

pujagani commented:
Thank you for helping take a look. The [LocalNode](https://github.com/SeleniumHQ/selenium/blob/6a8bf1ca7ed2c22eb736abb2ebb5cb58bc5d87bd/java/src/org/openqa/selenium/grid/node/local/LocalNode.java#L538-L539) logic, is for sending the result back to the user. The PR does something similar but to send to the driver. Mainly, if the sessionSlot is configured with custom capabilities, the PR merges that with incoming session capabilities and sends it to the driver factory.  My thought process was more around the session slot being the owner of the stereotype, the logic should sit there. LocaNode is more of a meditator and is aware of the session slots.  What do you think?

diemol commented on java/src/org/openqa/selenium/grid/node/local/SessionSlot.java:
```diff
@@ -144,7 +144,16 @@ public Either<WebDriverException, ActiveSession> apply(CreateSessionRequest sess
     try {
-      Either<WebDriverException, ActiveSession> possibleSession = factory.apply(sessionRequest);
+      Capabilities mergedCapabilities =
+        sessionRequest.getDesiredCapabilities().merge(this.getStereotype());
```
Would this overwrite the browser options passed by the user? Like, if the custom stereotype has a `goog:chromeOptions`, would it overwrite what the user has passed?

pujagani replied:
I think so. The capabilities of the slot override the capabilities of the incoming session request in case of custom driver-configuration.

diemol replied:
I think that is not ideal, because we would not honor the preferences set by the user. If we want to do this, probable this needs a deep merge of browser specific options. What do you think?

pujagani replied:
So we are trying to honour the preference settings on the Grid. My thinking was that if the Grid slots have preferences, they are meant to be universally applied to all incoming sessions that match, so ideally, they would take precedence.

In case of a deeper merge, what do you suggest we do differently that the current merge function?

diemol replied:
Right, but if a user wants to run headless, and the stereotype has a setting to set the browser language. Both things would live inside `goog:chromeOptions`. I think having the stereotype ones take precedence is unexpected and would break the behavior users currently have in Grid. 

The current merge logic replaces the value at the top level, if we want to do this, we should follow the same logic for browser preferences, like `goog:chromeOptions`, `moz:firefoxOptions`, etc...

pujagani replied:
Ahh okay, I think I see your point. Let me see how we can achieve that or what it will take to make it possible.

pujagani commented:
@diemol After giving your comments a thought, it made sense to move the logic to SessionCapabilitiesMutator since that is where we make some checks on the capabilities and add some capabilities as required. Additionally, as you correctly pointed out, I think it makes sense to give the user-sent capabilities priority over the stereotype capability to avoid any breaking changes. I have updated the changes accordingly and reverted the previous ones. Will appreciate your review. Thank you so much.

diemol commented on java/src/org/openqa/selenium/grid/node/config/SessionCapabilitiesMutator.java:
```diff
@@ -56,33 +57,137 @@ public Capabilities apply(Capabilities capabilities) {
+      // Merge browser specific stereotype and capabilities options
+      switch (browserName) {
+        case "chrome":
+        case "microsoftedge":
```
In some places in the Java code I see we consider `MicrosoftEdge` and `msedge`, maybe consider it here as well?

diemol commented on java/src/org/openqa/selenium/grid/node/config/SessionCapabilitiesMutator.java:
```diff
@@ -56,33 +57,137 @@ public Capabilities apply(Capabilities capabilities) {
+        toReturn.put("profile", rawProfile);
+      }
+
+      if (name.equals("binary")) {
+        Object binary = capsOptions.get("binary");
+        toReturn.put("binary", binary);
+      }
```
Same comment as the Chromium case.

pujagani replied:
I am afraid I don't understand. Which Chromium case is been referred to here?

diemol commented on java/src/org/openqa/selenium/grid/node/config/SessionCapabilitiesMutator.java:
```diff
@@ -56,33 +57,137 @@ public Capabilities apply(Capabilities capabilities) {
+        toReturn.put("profile", rawProfile);
+      }
+
+      if (name.equals("binary")) {
+        Object binary = capsOptions.get("binary");
+        toReturn.put("binary", binary);
+      }
+
+      if (name.equals("log")) {
+        Map<String, Object> logLevelMap =
+          (Map<String, Object>) capsOptions.get("log");
+        toReturn.put("log", logLevelMap);
+      }
+    }
```
Same as the `binary` case.

pujagani replied:
This is log-level property. For this, we should consider the user preference here right?

diemol replied:
Ah OK. Log level should be fine.

diemol reviewed (changes requested):
Thank you, I left a few comments.

diemol commented on java/src/org/openqa/selenium/grid/node/config/SessionCapabilitiesMutator.java:
```diff
@@ -56,33 +57,137 @@ public Capabilities apply(Capabilities capabilities) {
+        toReturn.put("extensions", stereotypeExtensions);
+      }
+
+      if (name.equals("binary")) {
+        Object binary = capsOptions.get("binary");
+        toReturn.put("binary", binary);
+      }
```
Somehow this comment was lost...

If I understand the code correctly, this would overwrite the `binary` location in the stereotype (if it exists), which is probably not ideal because the person who setup the Grid knows the browser paths.

pujagani replied:
Ahh okay, I understand now. Thank you for helping clarify this. I will fix it.

diemol reviewed (commented):
Added missing comment for the Chromium case.

diemol commented on java/src/org/openqa/selenium/grid/node/config/SessionCapabilitiesMutator.java:
```diff
@@ -126,9 +127,11 @@ private Map<String, Object> mergeChromiumOptions(Map<String, Object> stereotypeO
-      if (name.equals("binary")) {
-        Object binary = capsOptions.get("binary");
-        toReturn.put("binary", binary);
+      // Grid will know the browser paths so preference is given to stereotype value in this case
+      // If the value is not present in stereotype but present in capabilities only then use capabilities value
+      if (name.equals("binary") && stereotypeOptions.get("binary") == null) {
+          Object binary = capsOptions.get("binary");
+          toReturn.put("binary", binary);
```
I was thinking about this again and I think we should completely ignore `binary` when it comes in the session request capabilities. Ideally, the person who setup the Grid already has left all paths configured. 

This could also affect cloud vendors, probably BrowserStack sets the binary path where the browser is, IIRC Sauce does something like that. If a user sets the binary in a place where the browser is not present on the remote machine, session creation will fail.

pujagani replied:
I will make the changes. I see your point. Grid should be the source of truth in such cases.

diemol reviewed (approved):
Thank you, @pujagani!
